posthog/models/async_deletion/delete.py

```diff
@@ -0,0 +1,143 @@
+
+from posthog.client import sync_execute
+from posthog.models.async_deletion import AsyncDeletion, DeletionType
+from posthog.settings.data_stores import CLICKHOUSE_CLUSTER
+
+logger = structlog.get_logger(__name__)
+# Note: Session recording, dead letter queue, logs deletion will be handled by TTL
+TABLES_TO_DELETE_TEAMS_FROM = [
+    "person",
+    "person_distinct_id",
+    "person_distinct_id2",
+    "groups",
+    "cohortpeople",
+    "person_static_cohort",
+    "plugin_log_entries",
+]
+
+
+def run_event_table_deletions():
+    deletions = list(AsyncDeletion.objects.filter(delete_verified_at__isnull=True))
+
+    if len(deletions) == 0:
+        logger.debug("No AsyncDeletion to perform")
+        return
+
+    logger.info(
+        "Starting AsyncDeletion on `events` table in ClickHouse",
+        {"count": len(deletions), "team_ids": list(set(row.team_id for row in deletions)),},
+    )
+
+    conditions, args = _conditions(deletions)
+    sync_execute(
+        f"""
+        ALTER TABLE sharded_events
+        ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+        DELETE WHERE {" OR ".join(conditions)}
+        """,
+        args,
+    )
+
+    # Team data needs to be deleted from other models as well, groups/persons handles deletions on a schema level
+    team_deletions = [row for row in deletions if row.deletion_type == DeletionType.Team]
+
+    if len(team_deletions) == 0:
+        return
+
+    logger.info(
+        "Starting AsyncDeletion for teams on other tables",
+        {"count": len(team_deletions), "team_ids": list(set(row.team_id for row in deletions)),},
+    )
+    conditions, args = _conditions(team_deletions)
+    for table in TABLES_TO_DELETE_TEAMS_FROM:
+        sync_execute(
+            f"""
+            ALTER TABLE {table}
+            ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+            DELETE WHERE {" OR ".join(conditions)}
+            """,
+            args,
+        )
+
+
+def mark_deletions_done():
+    """
+    Checks and updates `delete_verified_at` for deletions
+    """
+    to_verify = []
+    unverified = _fetch_unverified_deletions_grouped()
+
+    for (deletion_type, _), async_deletions in unverified.items():
+        to_verify.extend(_verify_by_group(deletion_type, async_deletions))
+
+    if len(to_verify) > 0:
+        AsyncDeletion.objects.filter(pk__in=[row.pk for row in to_verify]).update(delete_verified_at=timezone.now())
+        logger.info(
+            "Updated `delete_verified_at` for AsyncDeletion",
+            {"count": len(to_verify), "team_ids": list(set(row.team_id for row in to_verify)),},
+        )
+
+
+def _fetch_unverified_deletions_grouped():
+    result = defaultdict(list)
+    for item in AsyncDeletion.objects.filter(delete_verified_at__isnull=True):
+        key = (item.deletion_type, item.group_type_index)
+        result[key].append(item)
+    return result
+
+
+def _verify_by_group(deletion_type: int, async_deletions: List[AsyncDeletion]) -> List[AsyncDeletion]:
```

Comment:
nit: I was briefly confused on a reference by the name of this function given group is also a deletion type